### PR TITLE
Feature: dkim for alternative domains

### LINKED ISSUE #1519: Sent emails from alternate domains don't support DKIM signing
Additionaly I'd love headers sent from alternate domains to only mention the alternate domain, with no mention of the "original" domain. My use is for alternate domains to just merge incoming emails.

I know postfix allows that, but do we want this feature available in mailu ?

I'd be happy with a workaround. For instance, the admin UI blocks creating domains and alternates with similar names, but I haven't tested yet what happens if I just alter tables in the db.

### PR DESCRIPTION
**Pull request by Jumper78**
## What type of PR?

feature

#### use same DKIM key of main domain for signing

Instead of dealing with key creation for each alternative domain, this implementation of the solution uses one key for all domains, the main domain and all alternative domains. Upon Rspamd requesting the DKIM key of a domain, it is not only checked if the domain is in the list of main domains, it also checked if it part of the alternative domains. If it is in this list, it sends the DKIM key of the connected main domain together with the name of the alternative domain.

#### show needed entries in the domain detailed view of the main domain

To make it easier for the admin to create the DKIM and DMARC entries (and the MX and SPF entries) for the alternative domains, we go through all alternative domains and print the entries.

### missing (and currently not planned to be added)

The zonefile at the top of the detail page will still only cover the main domain.

### Related issue(s)
- DKIM signing of the alternative domains is a requested feature; it closes #1519
- it keeps the original file based handling of DKIM keys; it does not implement #2952

## Prerequisites
Before we can consider review and merge, please make sure the following list is done and checked.
If an entry in not applicable, you can check it or remove it from the list.

- [ ] In case of feature or enhancement: documentation updated accordingly
- [x] Unless it's docs or a minor change: add [changelog](https://mailu.io/master/contributors/workflow.html#changelog) entry file.

### CODE REVIEW

**Comment by Jumper78** [comment]
screenshot-alternative-domain-details
This is how the part for the alternative domains looks like on the domain details page.

**Comment by nextgens** [comment]
The hunk at ``core/admin/mailu/internal/views/rspamd.py`` looks reasonable to me... and is all that's required to get the feature to work. I'm not sure we want the UI bits until the keys can be changed (which would make this a much bigger PR).

@Diman0 thoughts?

**Comment by Jumper78** [comment]
Thank you very much for the comment! Maybe a brief comment on what I thought when I wrote the UI change:
If the key for the alternative domain is the same as the key of the main domain, there is no way to change it separately. The "Regenerate Keys" button at the top of the page is sufficient. This is why (in my thinking) the UI change, which is only _showing_ the DNS entries for alternative domains, is adequate and should go with the change in `rspam.py`.
If you have in mind to have the alternative domain keys changeable separately from the other key, we should think of a completely different way to implement the feature "DKMI keys for alternative domains". Then the change in `rspam.py` should not be implemented at all.
I also thought about implementing separate keys for all alternative domains, so one individual one per domain. The implementation effort would be significantly higher, it would be more complicated to use DKMI and I currently see no advantage to have separate keys… except the fact that having one key per domain _feels_ like the “correct way” to implement it. In my opinion, if your server gets compromised, you will have to change all of the keys at the same time anyway and there is no advantage to have individual keys.
Let me know what you think about it.

**Comment by ser** [comment]
IMHO it's a reasonable solution to quickly move things forward, as currently it's impossible to send any mail from alternative domains due to DMARC enforcements on receiving sides.

**Review by nextgens — APPROVED** [review]
LGTM

**Comment by nextgens** [comment]
We can do it afterwards.

IMHO you want the ability to have separate keys to cover two main usecases:
- signing with more than one key (see #2050 for why that may be useful)
- allowing the import of keys from other systems (so that you don't have to wait for DNS propagation... or because you don't control the DNS zone of the said domain)

**Review by ghostwheel42 — APPROVED** [review]

**Comment by Aaron-Ritter** [comment]
> screenshot-alternative-domain-details This is how the part for the alternative domains looks like on the domain details page.

@Jumper78 @nextgens Is the display for Alternatives implemented? Do we have to regenerate the keys for it to work?

**Comment by nextgens** [comment]
@Aaron-Ritter  Do not highlight me without a good reason.

Yes and no.

There is no such convention outside of the BIND world; no one cares about the wire format. This PR does what is done for the main DKIM records; If you had suggestions to make the right time was before it got merged... and the right format was to submit a review.

**Comment by Aaron-Ritter** [comment]
> @Aaron-Ritter Do not highlight me without a good reason.
> 
> Yes and no.
> 
> There is no such convention outside of the BIND world; no one cares about the wire format. This PR does what is done for the main DKIM records; If you had suggestions to make the right time was before it got merged... and the right format was to submit a review.

I've highlighted you because i was not sure if it was actually released before creating an issue for it (done now) as the latest release notes where a copy from the previous one.

The [255 TXT limit](https://support.google.com/a/answer/11613097?hl=en#:~:text=For%20TXT%20records%20that%20include,with%20a%20255%2Dcharacter%20limit.) is something i noticed while testing and added it as a note but will create a issue accordingly.

**Comment by Jumper78** [comment]
Just for completeness (the question was already answered in the issue #3375): the code change was merged into the master branch. Since nearly everyone is currently still running 2024.06 branch releases (these are the latest releases at the point of writing), the feature is not visible unless you:
- either create the docker images on your own (see ...)
- or use the images created during this pull request (e.g. for the admin image you use `image: ${DOCKER_ORG:-ghcr.io/mailu}/${DOCKER_PREFIX:-}admin:pr-3350`)
- or you wait for the next major release, maybe something like 2024.XX

I explicitly do not recommend to combine the docker images from the 2024.06 releases with the `admin:pr-3350` docker image. It does work for me on my productive system ... but it is not a safe way to do it.

**Pull request merged**
